### PR TITLE
Algo Generic merge to master, other fixes

### PR DESCRIPTION
**Pull request by sujap**
Currently algo442 with depth 3 takes 1.3s, depth 4 takes 8.9s  with prune = (8,12)

### CODE REVIEW

File: chess/board.h [hunk 1]
```diff
@@ -26,21 +33,31 @@ namespace space {
 		Move(int v_sourceRank, int v_sourceFile, int v_destinationRank, int v_destinationFile, PieceType v_promotedPiece = PieceType::None) :
 			sourceRank(v_sourceRank), sourceFile(v_sourceFile), destinationRank(v_destinationRank), destinationFile(v_destinationFile), promotedPiece(v_promotedPiece)
 		{}
-		inline std::string as_string() const {
+		Move(const std::string &s);
```
Comment (appu226) [inline chess/board.h:36]: I am thinking of converting some of these types to "Plain old data" types (https://en.cppreference.com/w/cpp/named_req/PODType), so at some point i will convert these constructors to static factory functions.
Just FYI... for now it's perfectly fine.

File: chess/board.h [hunk 2]
```diff
@@ -26,21 +33,31 @@ namespace space {
...
+  
 	struct Piece {
 		PieceType pieceType;
 		Color color;
-		char as_char() const;
-	};
-	class IBoard {
+		Piece() {}
+		Piece(PieceType _p, Color _c) : pieceType(_p), color(_c) {}
+		Piece(char c);
```
Comment (appu226) [inline chess/board.h:55]: Umm... By adding these constructors, you are adding a performance penalty, since this is no longer a plain old data type.
Consider removing these constructors.
https://en.cppreference.com/w/cpp/named_req/PODType

File: chess/board_impl.h [hunk 3]
```diff
@@ -51,6 +54,9 @@ namespace space {
 		bool m_canBlackCastleRight;
 		Color m_whoPlaysNext;
 		bool checkObstructions(Move m) const;
+		bool canMove(Move m) const;
+		bool checkPathEmpty(Move m) const;
```
Comment (appu226) [inline chess/board_impl.h:58]: Just for my curiosity: Why did you add these functions? They are not used anywhere, are they?

Comment (sujap) [reply chess/board_impl.h:58]: Added these two for alternate implementation of isUnderCheck, also can be useful for identifying attacks & defenses (not implemented yet)

**Review by appu226 — APPROVED** [review]
